### PR TITLE
Fix #7704: Launching coqide through PATH fails.

### LINKED ISSUE #7704: Launching `coqide` through `PATH` fails.
#### Version

Coq master 1136dde7b523047e6091d6e6decb45183e42fc21

#### Operating system

Linux

#### Description of the problem

When launching CoqIDE using a path that is not relative to the current directory nor absolute, the worker fails to connect back in 10s, and I get an error message saying Coq is dead.

Example:

```
/absolute/path/to/coqide
```

works

```
bin/coqide
```

when current directory is the Coq root, also works, but:

```
coqide
```

when `coqide` is in `PATH` typically fails.

I guess some wrong assumptions were done in some path handling code. This behavior seems like a regression, I'm pretty sure it used to work.

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Review by ejgallego — CHANGES_REQUESTED** [review]
Please document this issue properly, in most cases `exe_name` is set to `argv[0]` in `startup.c`, so why is that not enough?

**Comment by ejgallego** [comment]
To be clear I am not asking for a code change but for a better commit message.

**Comment by SkySkimmer** [comment]
Updated (and added test).

**Review by gares — DISMISSED** [review]

**Comment by Zimmi48** [comment]
Emilio's request was about the commit message, which has been improved.

File: test-suite/Makefile [hunk 1]
```diff
@@ -519,6 +519,7 @@ $(patsubst %.sh,%.log,$(wildcard misc/*.sh)): %.log: %.sh $(PREREQUISITELOG)
 	@echo "TEST      $<"
 	$(HIDE){ \
 	  echo $(call log_intro,$<); \
+	  export COQBIN="$(BIN)"; \
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply test-suite/Makefile]: Changed for just BIN.

**Comment by ejgallego** [comment]
Umm, still something is puzzling me, I am not sure this is not a bug in OCaml. Let's take this program:

```ocaml
let _ =
  Format.printf "%s,%s\n%!" Sys.argv.(0) Sys.executable_name
```
Adding `k` to my path and using `ocamlc`:
```
/home/egallego/my/lib/shell/k,/home/egallego/my/lib/shell/k
```
however using `ocamlopt`:
```
k,/home/egallego/my/lib/shell/k
```
gives a different result. Ummm, I am not sure I like this discrepancy here, even if I understand the reason behind it.

Anyways I shall merge I think but IMHO we should document this ocaml problem better.

**Comment by ejgallego** [comment]
Another question I do have is why do we need a normalized argument name for a path call.

As far as I can see, if the user calls `coqtop` from path, `coqidetop` for example should be also searched using PATH, right?

What is the problem we are having here? Why is coqtop in path but not the rest of the proper toplevels?

The original implementaton indeed tried to rely on path for that; I guess using the canonical one is not a problem but I'd like to understand what the problem is in the first place.

**Comment by SkySkimmer** [comment]
Here's a log of launching coqide through PATH with debug printing
~~~
(Spawn  ,21597) EXEC: coqidetop -main-channel stdfds --xml_format=Ppcmds -async-proofs on
(Spawn  ,21597) PID 21604
Welcome to Coq debian-hibiki:/home/gaetan/dev/coq/coq,def-const-nosubst (efb27a1bbdda4d4b36ece4e6fa4359bfeddc9549)
(Spawn  ,21604) EXEC: ./coqproofworker.opt -main-channel stdfds --xml_format=Ppcmds -worker-id proofworker:0 -async-proofs-worker-priority low
(Spawn  ,21604) PID 21607
Handshake with proofworker:0 failed: End_of_file
(Spawn  ,21597) EXEC: coqidetop -main-channel stdfds --xml_format=Ppcmds -async-proofs on
(Spawn  ,21597) PID 21609
(Spawn  ,21597) This process is already dead
Welcome to Coq debian-hibiki:/home/gaetan/dev/coq/coq,def-const-nosubst (efb27a1bbdda4d4b36ece4e6fa4359bfeddc9549)
(Spawn  ,21609) EXEC: ./coqproofworker.opt -main-channel stdfds --xml_format=Ppcmds -worker-id proofworker:0 -async-proofs-worker-priority low
(Spawn  ,21609) PID 21613
Handshake with proofworker:0 failed: End_of_file
~~~
When argv.(0) is unqualified the dirname in System.get_toplevel_path returns ./ which is not what we want.
coqidetop gets launched correctly because https://github.com/coq/coq/blob/a1fc621b943dbf904705dc88ed27c26daf4c5e72/ide/ideutils.ml#L295-L304 falls back to PATH when the file doesn't exist, but for the proofworker there is no such fallback so it calls spawn with ./coqproofworker.opt (as we can see in the log).

**Comment by SkySkimmer** [comment]
>Ummm, I am not sure I like this discrepancy here, even if I understand the reason behind it.

What is the reason?

**Comment by SkySkimmer** [comment]
>When argv.(0) is unqualified the dirname in System.get_toplevel_path returns ./ which is not what we want.

We could handle that on our side but using `executable_name` does it for us so let's go with it.

**Comment by ejgallego** [comment]
Thanks for the explanation.

> What is the reason?

The reason for the discrepancy is that for bytecode programs, `argv.(0)` is `ocamlrun` so the runtime tries some normalization. But now I see what the problem is:

> When argv.(0) is unqualified the dirname in System.get_toplevel_path returns ./ which is not what we want.
> coqidetop gets launched correctly because

Oh that's indeed a bug in the code, but I am not sure if we should use the canonical name or just check the name better and still allow for the path to override the install. Let me think for a minute.

**Comment by ejgallego** [comment]
> Let me think for a minute.

So WDYT @SkySkimmer ? Should we allow PATH to override file-path based resolution [and thus modify the code], or should we take the canonical, but less flexible approach?

The spirit of the buggy code was 1) , anyways, IMHO the code needs a comment.

**Comment by maximedenes** [comment]
This is a very serious bug. Can we merge this?

If this really requires more work before being merged (I doubt it), I suggest we revert the commit that introduced this regression.

**Comment by ejgallego** [comment]
> Canonical. Feel free to add a comment.

We will indeed have to modify the whole setup, but let's merge for now.

**Review by ejgallego — APPROVED** [review]

**Comment by ejgallego** [comment]
Indeed we need to further tweak the strategy here as this creates problems when Coq is installed using symlinks, such as in Dune builds [which seems essential to save disk space]

I think the strategy here should be way simpler, take `argv.[0]` and extract the directory, then append the proper subbinary name.

**Pull request merged**